### DISCUSSION TITLE
Far как консоль для ssh подключений

### DISCUSSION THREAD

**Discussion by ZEV1416** [elfmz/far2l#1291] [Ideas]
Суть проблемы: при подключении по ssh, даже с -X, не работают нормально быстрые клавиша. Текстовым редактором на ноутбуке становится пользоваться невозможно, т.к. Ctrl-C не работает, а Ctrl-Ins нажать на неполной клавиатуре - проблематично.
Выход: запускаем из под Far2l ssh, и тогда работает всё как надо.
Но - у меня подключений по ssh несколько десятков, для них созданы ярлыки, меню, и т.п. - удобно. 
И вот нельзя запустить Far2l с неким параметром, в котором передать команду, чтобы Far2l ее просто выполнил, а потом закрылся.
Просьба - добавьте такой ключ? 
Тогда можно будет сделать ярлык на команду:
far2l -e ssh user@ip.address и получить подключение к нужному серверу прямо из far2l, используя его по сути как консоль.
Этакая замена форка putty для far2l под linux...

**Comment by elfmz** [comment]
Вы можете в NetRocks создать подключения sftp или scp. Каждое такое подключение по совместительствую еще и ssh команды исполнять умеет. То есть потом заходите в NetRocks в подключение, пишете в ком строке far2l - и вуаля.

**Reply by unxed** [reply]
Как я понял, речь шла о том, чтобы при старте far2l с определенными параметрами сразу поднимать определенное соединение sftp в нетроксе. Вроде же такое можно было? Только не могу вспомнить, как. Макросом на автозапуск?

**Reply by ZEV1416** [reply]
Из предложенных по --help ключей можно только выполнить команду и передать ее результат редактору или в просмотрщик. Первое - то, что мне надо (выполнить команду). А вот затем - нужно просто закрыться. Открывать редактор/просмотр уже не надо. А такого ключа нет. :-)

**Reply by shmuz** [reply]
Если установлен пакет LuaFAR-2L, то (поскольку Фар может принять при запуске две команды) можно указать в качестве второй команды `lfe:-eactl.Quit()`.

**Comment by shmuz** [comment]
С NetRocks я пока не знаком, но если он поддерживает команду подключения через префикс командной строки, то эта команда может быть одним из параметров запуска Фара.

**Reply by elfmz** [reply]
действительно, работает, но открывается только подключение, дальше надо в нем вручную натайпать команду (far2l в данном случае): `far2l 'net:<sftp:user@192.168.1.11>'` - где между треугольными скобочками имя подключения как оно видно в тайтле панели если просто руками зайти в подключение NetRocks-а

**Reply by shmuz** [reply]
Вот пример для Фар3. В ярлыке на рабочем столе указана команда
`<path-to>\Far.exe sftp://demo:password@test.rebex.net:22`
Если дабл-кликнуть на этот ярлык, запускается Фар и автоматически происходит соединение. Ничего не надо вводить вручную.

**Reply by ZEV1416** [reply]
Получился какой-то набор костылей.
1. Из под Linux подключение проходит, но приходится вводить пароль, хотя есть ключи авторизации для ssh.
2. После подключения приходится писать far2l, чтобы запустился удаленный far.
3. После выхода из удаленного far я оказываюсь в локальном, из которого приходится еще раз выходить.
А весь смысл - только чтобы нормально работали быстрые клавиши и буфер обмена на удаленном сервере. 
Проделать такое в случае, если нужно что-то долго делать на удаленном сервере и хочется комфорта - можно. Но повседневно, чтобы "забежать на секунду" - слишком много лишних телодвижений...
А из-за невозможности работы в текстовом редакторе при обычном входе через ssh - фар становится неудобным, и отходит на второй план. Проще уж запустить mc, там работа с блоками текста на функциональных клавишах, и работает "из коробки"...
Пока подожду, вдруг будут улучшения...

**Reply by unxed** [reply]
Мне, кстати, тоже не хватает возможности запустить удаленный far2l по ssh в один клик.

**Reply by elfmz** [reply]
1 насчет ключей, а в настройках подключения указано использовать SSH ключ? Protocol options/Authentification/Private key?

**Reply by ZEV1416** [reply]
Каюсь, было не указано. Сейчас всё в порядке :-)

**Comment by dmikushin** [comment]
Ребята, посмотрите, пожалуйста, как сделано удалённое подключение в Midnight Commander и вдохновитесь его простотой: пункты меню F9 открывают строку подключения, которая умеет брать конфигурации из `~/.ssh/config`. Подключение переводит одну панель в режим просмотра удалённого сервера. Можно открыть второй удалённый сервер на другой панели.

![Screenshot from 2024-12-31 01-52-31](https://github.com/user-attachments/assets/c4ae9d4d-0601-4c8d-993a-95b8fd028572)
![Screenshot from 2024-12-31 01-54-06](https://github.com/user-attachments/assets/3ed23eff-f641-4583-8ab2-f49a066aa105)
![Screenshot from 2024-12-31 01-56-10](https://github.com/user-attachments/assets/85d95a2e-8365-4ff5-9869-1894a0f6120a)

Не поймите меня неправильно, я очень рад увидеть FAR в Linux! Но эргономика и UX тоже значат очень многое для успеха.

**Reply by akruphi** [reply]
> очень рад увидеть FAR в Linux! 

Спасибо большое за добрые слова.

> Можно открыть второй удалённый сервер на другой панели.

Это внутри far2l через NetRocks сейчас прекрасно работает.

> Но эргономика и UX тоже значат очень многое для успеха

Эргономика и UX far2l старается максимально повторять windows прародителя.
Соединения реализуются плагином NetRocks, доступ через **F9**->Left/Right->Location menu->NetRocks или быстро по **Alt+F1**/**Alt+F2** ->NetRocks.

![image](https://github.com/user-attachments/assets/ea1fae28-3d69-47da-92b0-16f67330500f)
![image](https://github.com/user-attachments/assets/375e7dd8-fc92-4d24-9dc2-4ad83499fc7b)

В каком соотношении Вас напрягают
- лишний шаг внутри Location menu по выбору NetRocks
- неудобство дальнейшего создания соединения внутри NetRocks ?

Что на этих этапах и как именно улучшить?

**Reply by dmikushin** [reply]
Спасибо за совет! Да, Alt+F2 повторяет поведение Norton Commander для выбора drive. Логично что для Windows NetRocks находится там. В файловой системе UNIX драйвы не имеют смысла, поэтому меню заполнено точками монтирования. Но в списоке монтирования трудно отделить важное от "мусора", например у меня в него попали какие-то слои докера (запускаю far2l на хосте):

![Screenshot from 2024-12-31 14-10-33](https://github.com/user-attachments/assets/435524c1-8ed1-425b-b16f-086af3525901)

Похоже, NetRocks не может разыменовать имя хоста с помощью `~/.ssh/config`, но наличие меню OpenSSH config files предполагает, что должен. Выглядит сломанным, а без `~/.ssh/config` можно управлять только очень простыми тоннелями.

![Screenshot from 2024-12-31 13-30-28](https://github.com/user-attachments/assets/c08d8d37-44fd-4268-b1d7-4924fc591dce)

> Что на этих этапах и как именно улучшить?

Если бы я умел, то первым делом сделал бы нативный shell юзера в оболочке Ctrl+O. Шеллы типа zsh и fish сейчас настолько продвинутые и привычные, что игнорировать их поддержку просто никак нельзя. Вторым шагом реализовал бы плагин для размещения кастомных элементов меню в F9->Left/Right и закинул бы SSH/SFTP/SMB прямо туда по умолчанию для Linux.

**Reply by akruphi** [reply]
> в списоке монтирования трудно отделить важное от "мусора", например у меня в него попали какие-то слои докера

В этом списке, нажав **F9**, можно перейти к настройкам, где либо выключить вообще показ точек монтирования, либо настроить маски исключений для отсечения только не нужных точек монтирования.
![image](https://github.com/user-attachments/assets/a61c1cbe-a68a-42c4-abef-aaa07098bc0f)

> Похоже, NetRocks не может разыменовать имя хоста с помощью ~/.ssh/config

Должно работать. Интересно какая у вас версия far2l и какую версию libssh он использует и что из навороченного в `~/.ssh/config` не понимается.

> сделал бы нативный shell юзера в оболочке Ctrl+O.

Настройка для выбора shell есть в **F9**->Options->Command line settings, но пока полноценно это пока не работает - см. пояснение в https://github.com/elfmz/far2l/issues/1264#issuecomment-1045334082

**Reply by dmikushin** [reply]
> Должно работать. Интересно какая у вас версия far2l и какую версию libssh он использует и что из навороченного в ~/.ssh/config не понимается.

Использую версию 3b4ff5b6bf1091412d2b27ad9ad3fcfa9e028618

В `~/.ssh/config` есть такая запись:

```
Host gambetta
      User marcusmae
      Hostname localhost
      Port 64070
      IdentityFile ~/.ssh/keys/id_ed25519.gambetta
      IdentitiesOnly yes
      ForwardAgent yes
      ProxyJump oracle
```

Соотвественно, в NetRocks я просто пишу gambetta и получаю:

![Screenshot from 2024-12-31 20-36-22](https://github.com/user-attachments/assets/da29ec2c-9c42-48bf-8b10-2ba264e8e4e1)

Версия libssh - из пакетов Ubuntu 22.04:

![Screenshot from 2024-12-31 20-52-47](https://github.com/user-attachments/assets/c06b3436-dcaf-41da-b996-a3c1907c954a)

**Reply by akruphi** [reply]
В терминале установите переменные для логирования согласно https://github.com/elfmz/far2l/wiki/Running-far2l-with-debug-logging-enabled и из этого терминала запустите far2l. Возможно это поможет отловить почему `~/.ssh/config` не отрабатывается в вашем случае.
